### PR TITLE
Improve command output for duplicate and near duplicate appends + test cases

### LINKED ISSUE #206: [PE-D] No error message thrown when attempting to add duplicate asset/policies/liabilities
It is good that I am unable to actually add the same asset, however there is no error message thrown. It still says new policy/asset/liability added (so I would expect the number of entries to go up) even though the number of entries remain the same in the table.

-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: yongxiangng/ped#4

### LINKED ISSUE #207: [PE-D] Able to add asset/policies/liabilities of the same name
Although this is quite similar to one of the issues I've opened, I don't think it's exactly the same. It seems that the team wants to handle the case to prevent duplicate entries (putting the exact same command doesn't create 2 entries), however I am able to add assets with the same asset name but different value. As such, it seems counter intuitive to me that this is allowed but the other isn't

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: yongxiangng/ped#5

### PR DESCRIPTION
**Pull request by DonnyS57**
Fixes #206.
Fixes #207

Added a few negative test cases for AppendCommand & RemoveCommand as well

### CODE REVIEW

File: src/main/java/donnafin/logic/commands/AppendCommand.java [hunk 1]
```diff
@@ -54,12 +72,26 @@
...
+        requireNonNull(personAdapter);
+        requireNonNull(policy);
         this.personAdapter = personAdapter;
         this.hashableNewValue = policy;
         this.editor = pa -> {
             Set<Policy> policies = new HashSet<>(pa.getSubject().getPolicies());
-            policies.add(policy);
-            pa.editPolicies(policies);
+            boolean containsSimilarName = policies.stream().anyMatch(item ->
+                    item.getName().equalsIgnoreCase(policy.getName()));
```
Comment (bharathcs) [inline src/main/java/donnafin/logic/commands/AppendCommand.java:82]: tbh why don't you use `.anyMatch(item -> item.isWeakDuplicate(policy))`. Same could be used for assets and liabilities. If not tbh, you can remove the `isWeakDuplicate` function from all the **attribute** classes.

Comment (bharathcs) [reply src/main/java/donnafin/logic/commands/AppendCommand.java:82]: Feel free to update the `isWeakDuplicate` methods for policies, assets and liabilities. You may also wanna do a directory wide search for '@Donny' btw.

File: src/main/java/donnafin/logic/commands/AppendCommand.java [hunk 2]
```diff
@@ -54,12 +72,26 @@
...
+            if (policies.contains(policy)) {
+                try {
+                    messageRESULT = MESSAGE_DUPLICATE_POLICY;
+                    throw new CommandException(MESSAGE_DUPLICATE_POLICY);
+                } catch (CommandException e) {
+                    e.printStackTrace();
+                }
```
Comment (bharathcs) [inline src/main/java/donnafin/logic/commands/AppendCommand.java:89]: Hmm, why are we throwing a command exception only to swallow it with a printstactrace? I don't quite get the code here.

File: src/test/java/donnafin/logic/commands/AppendCommandTest.java [hunk 3]
```diff
@@ -61,6 +67,14 @@ public void setUp() throws IllegalValueException {
...
         testAsset = new Asset("Good Class Bungalow", "Property",
                 "$2000000", "Paid in full. No debt.");
+        similarAsset = new Asset("Good Class Bungalow", "Property",
+                "$20000000", "Paid in full. No debt.");
+        similarLiability = new Liability("Mortgage debt", "Debt", "$200000",
+                "23 year loan from DBS Bank.");
+        similarPolicy = new Policy("Golden Mile", "AIA",
+                "$1800", "$100", "$200");
+
+
```
Comment (bharathcs) [inline src/test/java/donnafin/logic/commands/AppendCommandTest.java:77]:
```suggestion
```

File: src/test/java/donnafin/logic/commands/RemoveCommandTest.java [hunk 4]
```diff
@@ -127,5 +128,12 @@ public void removePolicy_changesPolicyList() throws CommandException, ParseExcep
         assertEquals(GEORGE, personAdapter.getSubject());
     }
 
+    @Test
+    public void execute_invalidIndex_throwsCommandException() throws ParseException {
+        RemoveCommand testCommand = new RemoveCommand(personAdapter, fieldLiability, index2);
+        assertCommandFailure(testCommand, model, "No such index found.");
+    }
+
+
```
Comment (bharathcs) [inline src/test/java/donnafin/logic/commands/RemoveCommandTest.java:137]:
```suggestion
```

**Review by bharathcs — COMMENTED** [review]
Nice work! Just some points regarding `AppendCommand` policy constructors, but the same comments can be applied to the liabilities and assets too.

File: src/main/java/donnafin/logic/commands/AppendCommand.java (hunk 1 above)
Comment (DonnyS57) [reply src/main/java/donnafin/logic/commands/AppendCommand.java:82]: Do you mean that there is already a ```isWeakDuplicate``` function or do you want me to make one?

File: src/main/java/donnafin/logic/commands/AppendCommand.java (hunk 2 above)
Comment (DonnyS57) [reply src/main/java/donnafin/logic/commands/AppendCommand.java:89]: Hmmm I'm not sure if pa (the lambda function) can throw an exception cause when I delete the try catch block it'll just show up as an error

File: src/main/java/donnafin/logic/commands/AppendCommand.java (hunk 1 above)
Comment (bharathcs) [reply src/main/java/donnafin/logic/commands/AppendCommand.java:82]: Oops I guess the function was named isPossibleDuplicate. Rip should that be renamed as isWeakDuplicates?

File: src/main/java/donnafin/logic/commands/AppendCommand.java (hunk 2 above)
Comment (bharathcs) [reply src/main/java/donnafin/logic/commands/AppendCommand.java:89]: Oh then why the throw new CE at all?

Comment (DonnyS57) [reply src/main/java/donnafin/logic/commands/AppendCommand.java]: alright, I deleted the CE!

File: src/test/java/donnafin/logic/commands/AppendCommandTest.java [hunk 5]
```diff
@@ -121,4 +133,44 @@ public void appendNewAsset_changesAssetList() throws CommandException {
 
         assertNotEquals(GEORGE, personAdapter.getSubject());
     }
+
+    @Test
+    public void appendNewAsset_similarAssetName() throws CommandException {
+        AppendCommand testCommand = new AppendCommand(personAdapter, testAsset);
+        AppendCommand testCommand2 = new AppendCommand(personAdapter, similarAsset);
+        testCommand.execute(model);
+        assertCommandSuccess(testCommand2, model, "New asset added. "
```
Comment (bharathcs) [inline src/test/java/donnafin/logic/commands/AppendCommandTest.java:142]: Could you switch out the magic literal for a reference to the possible duplicate asset String constant.

File: src/main/java/donnafin/logic/commands/AppendCommand.java [hunk 6]
```diff
@@ -84,12 +121,21 @@ public AppendCommand(PersonAdapter personAdapter, Liability liability) {
...
-            assets.add(asset);
-            pa.editAssets(assets);
+            boolean containsSimilarName = assets.stream().anyMatch(item ->
+                    item.isPossibleDuplicate(asset));
+            if (assets.contains(asset)) {
+                messageRESULT = MESSAGE_DUPLICATE_ASSET;
+            } else {
+                assets.add(asset);
+                pa.editAssets(assets);
+                messageRESULT = containsSimilarName ? MESSAGE_SIMILAR_ASSET : MESSAGE_SUCCESS_ASSET;
```
Comment (bharathcs) [inline src/main/java/donnafin/logic/commands/AppendCommand.java:137]: Cases for messageResult a bit wierd, here and elsewhere

File: src/test/java/donnafin/logic/commands/AppendCommandTest.java (hunk 5 above)
Comment (bharathcs) [reply src/test/java/donnafin/logic/commands/AppendCommandTest.java:142]: Here and other places in the test files

**Review by bharathcs — CHANGES_REQUESTED** [review]
Hey there, just some more changes for code quality.

File: src/test/java/donnafin/logic/commands/AppendCommandTest.java (hunk 5 above)
Comment (DonnyS57) [reply src/test/java/donnafin/logic/commands/AppendCommandTest.java:142]: alright on it!

**Review by bharathcs — APPROVED** [review]
Nice!

**Pull request merged**